### PR TITLE
Support load from Bazel

### PR DESCRIPTION
**Pull request by liuliu**
This added support to load Swift libraries from a Bazel-based monorepo.
It makes assumption that Jupyter is launched from within that monorepo.

How to:

```
%bazel "@PythonKit//:PythonKit"
```

This will install Bazel packages available locally, and made them
available for the notebook to use.

There is a more detailed discussion of steps we do to achieve this.

---

I am not sure if this fits to upstream, since it is substantial of work and
many in the Swift community may not use Bazel as their build tools.
nevertheless, putting this up to receive comments. I am OK with not
upstreaming this, or other suggestions (like modularize the shared parts
between Bazel / SwiftPM, and upstreaming these refactorings while keeping
Bazel portion to myself etc).

### CODE REVIEW

**Comment by marcrasi** [comment]
Wow, looks cool! Sorry I didn't notice this earlier, I was working hard to meet a paper deadline. I'll try to take a look today :)

File: swift_kernel.py [hunk 1]
```diff
@@ -823,6 +837,184 @@ def is_valid_dependency(path):
...
+
+        self.send_response(self.iopub_socket, 'stream', {
+            'name': 'stdout',
+            'text': 'Installation complete!\n'
+        })
+        # Switch back to whatever we previous from.
+        os.chdir(pwd)
+        self.already_installed_packages = True
+
+
```
Comment (marcrasi) [inline swift_kernel.py:1017]: (Optional) This `SwiftKernel` megaclass is becoming pretty unweildy and it's getting hard to keep track of what methods are used for what functionality. Splitting the installation functionality into a different class in a different file could be a good place to start fixing this :)

**Review by marcrasi — COMMENTED** [review]
This looks pretty good to me.

I'd be happy to take this upstream and devote minimal effort to maintain it (simple fixes, update function calls when I refactor stuff, etc). If it becomes difficult to maintain, then I'll turn it off and ask you to fix it. If that sounds better for you than maintaining a separate copy, then let's upstream it!

I would like a test before accepting it upstream, so that it is easy for me to notice when it breaks. [This](https://github.com/google/swift-jupyter/blob/master/test/tests/simple_notebook_tests.py) has some simple examples you can follow. You can modify the [Dockerfile](https://github.com/google/swift-jupyter/blob/master/docker/Dockerfile) to install any dependencies that are required for testing this. The bottom of the readme has instructions how to run the tests in docker.

**Comment by liuliu** [comment]
Thanks! Sounds good. I will do some clean up and add Bazel to the Dockerfile this weekend.

**Review by liuliu — COMMENTED** [review]
Did suggested changes in `swift_kernel.py`

Make sure the Dockerfile can build and launch. In the launched image, you should be able to:
```
%bazel "@PythonKit//:PythonKit"
```
to load Bazel built PythonKit.

Have one question about the placement of Dockerfile.

**Comment by liuliu** [comment]
@marcrasi friendly ping :)

**Comment by liuliu** [comment]
My changes have diverged too much from the original, going to fork `swift-jupyter` and close this PR.

**Comment by philipturner** [comment]
Referencing https://github.com/google/swift-jupyter/pull/121#discussion_r526442035:

With Swift-Colab 2.0, it should be possible to install both SwiftPM and Bazel dependencies at the same time. As long as I can install Bazel itself at runtime, and build the Bazel packages at runtime through shell commands, it should be doable. Is there a reliable way to locate the `%.swiftmodule` and `module.modulemap` files produced by Bazel?

@liuliu I know that you created a Swift library that builds using Bazel instead of SwiftPM. It would be great to see if we can get that running in Colab.

**Comment by liuliu** [comment]
> With Swift-Colab 2.0, it should be possible to install both SwiftPM and Bazel dependencies at the same time. As long as I can install Bazel itself at runtime, and build the Bazel packages at runtime through shell commands, it should be doable. Is there a reliable way to locate the `%.swiftmodule` and `module.modulemap` files produced by Bazel?

Getting `%.swiftmodule` and `module.modulemap` is actually straightforward with Bazel, you can make `aquery` to get the command that executed and extract the module path directly there:
For `.swiftmodule`: https://github.com/liuliu/swift-jupyter/blob/main/lab/kernels/swift/swift_kernel.py#L1215
For `module.modulemap`: https://github.com/liuliu/swift-jupyter/blob/main/lab/kernels/swift/swift_kernel.py#L1285

The more involved part is to general the proper `.so` library, because AFAICT `rules_swift` doesn't support that and do it naively, you will have all symbols stripped. That's why I included this new rule: https://github.com/liuliu/swift-jupyter/blob/main/lab/kernels/swift/whole_archive.bzl and use it to generate the `.so` library like SwiftPM does: https://github.com/liuliu/swift-jupyter/blob/main/lab/kernels/swift/swift_kernel.py#L1150

**Comment by philipturner** [comment]
I look forward to implementing this in Swift-Colab v2.3!

**Pull request closed without merging**